File: Atcoder_#341/ruby/b_foreign_exchange.rb
# 入力を受け取る
n = gets.to_i
a = gets.split(" ").map(&:to_i)

s = []
t = []

(n - 1).times do
  si, ti = gets.split.map(&:to_i)
  s << si
  t << ti
end

# i=1から順にs, tの処理を行っていく
def max_currency_n(n, a, s, t)
  (0...n-1).each do |i|
    if a[i] >= s[i]
      # 最大限通貨を次の国に移動させる
      moves = a[i] / s[i]
      a[i] -= moves * s[i]
      a[i + 1] += moves * t[i]
    end
  end
  a[n-1]
end

result_last_n = max_currency_n(n, a, s, t)
puts result_last_n

File: Atcoder_#341/javascript/b_foreign_exchange.js
function main(input) {
    const lines = input.trim().split('\n');
    const n = parseInt(lines[0], 10);
    const a = lines[1].split(' ').map(Number);
  
    const s = [];
    const t = [];
  
    for (let i = 0; i < n - 1; i++) {
      const [si, ti] = lines[i + 2].split(' ').map(Number);
      s.push(si);
      t.push(ti);
    }
  
    // i=1から順にs, tの処理を行っていく
    function maxCurrencyN(n, a, s, t) {
      for (let i = 0; i < n - 1; i++) {
        if (a[i] >= s[i]) {
          const moves = Math.floor(a[i] / s[i]);
          a[i] -= moves * s[i];
          a[i + 1] += moves * t[i];
        }
      }
      return a[n - 1];
    }
  
    const result = maxCurrencyN(n, a, s, t);
    console.log(result);
  }
  
  const fs = require('fs');
  const input = fs.readFileSync('/dev/stdin', 'utf8');
  main(input);
